File: xtask/src/conventional_commit.rs
use enum_iterator::{Sequence, all};
use git2::Commit;
use std::fmt::Formatter;

#[derive(Debug, PartialEq, Sequence)]
pub enum ConventionalCommitKind {
  Feature,
  Fix,
  Refactor,
  Performance,
  Test,
  Style,
  Docs,
  Build,
  Operations,
  Chore,
}

impl std::fmt::Display for ConventionalCommitKind {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    let str = match self {
      Self::Feature => "feat",
      Self::Fix => "fix",
      Self::Refactor => "refactor",
      Self::Performance => "perf",
      Self::Test => "test",
      Self::Style => "style",
      Self::Docs => "docs",
      Self::Build => "build",
      Self::Operations => "ops",
      Self::Chore => "chore",
    };
    write!(f, "{}", str)
  }
}

#[derive(Debug)]
pub struct ConventionalCommit {
  #[allow(dead_code)]
  pub sha: String,
  pub kind: ConventionalCommitKind,
  pub scopes: Vec<String>,
  pub is_breaking: bool,
  pub summary: String,
  #[allow(dead_code)]
  pub body: Option<String>,
}

impl ConventionalCommit {
  pub fn parse(sha: &str, message: &str) -> Result<Self, crate::Error> {
    let lines = message.lines().collect::<Vec<_>>();
    let body = lines.get(2..).map(|x| x.join("\n"));
    let re =
      regex::Regex::new(r"^(?P<kind>\w+)(?:\((?P<scopes>.+)\))?(?P<breaking>!)?: (?P<summary>.+)$")
        .unwrap();
    let caps = re
      .captures(
        lines
          .first()
          .ok_or(crate::Error::ConventionalCommitParseError)?,
      )
      .ok_or(crate::Error::ConventionalCommitParseError)?;
    let kind_val = caps
      .name("kind")
      .ok_or(crate::Error::ConventionalCommitParseError)?;
    let kind = all::<ConventionalCommitKind>()
      .find(|x| x.to_string() == kind_val.as_str())
      .ok_or(crate::Error::ConventionalCommitParseError)?;
    let summary = caps
      .name("summary")
      .ok_or(crate::Error::ConventionalCommitParseError)?
      .as_str()
      .to_string();
    let scopes = caps
      .name("scopes")
      .map(|x| {
        x.as_str()
          .split(',')
          .map(|x| x.trim().to_string())
          .collect::<Vec<_>>()
      })
      .unwrap_or_default();
    let is_breaking = caps.name("breaking").is_some() || message.contains("BREAKING CHANGE:");
    Ok(Self {
      sha: sha.to_string(),
      kind,
      scopes,
      is_breaking,
      summary,
      body,
    })
  }
}

impl<'a> TryFrom<&Commit<'a>> for ConventionalCommit {
  type Error = crate::Error;

  fn try_from(value: &Commit<'a>) -> Result<Self, Self::Error> {
    let sha = value.id().to_string();
    let message = value
      .message()
      .ok_or(crate::Error::ConventionalCommitParseError)?;
    Self::parse(&sha, message)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn parse() {
    let commit = ConventionalCommit::parse(
      "123",
      r#"feat(core): implement `my-module` crate

This commit implements `my-module` crate.
"#,
    );
    assert!(commit.is_ok());
    let commit = commit.unwrap();
    assert_eq!(commit.sha, "123");
    assert_eq!(commit.kind, ConventionalCommitKind::Feature);
    assert_eq!(commit.scopes, vec!["core"]);
    assert!(!commit.is_breaking);
    assert_eq!(commit.summary, "implement `my-module` crate");
    assert_eq!(
      commit.body.unwrap(),
      "This commit implements `my-module` crate."
    );
  }

  #[test]
  fn multiple_scopes() {
    let commit = ConventionalCommit::parse("123", r#"feat(core, cli): hotfix"#).unwrap();
    assert_eq!(commit.scopes, vec!["core", "cli"]);
  }

  #[test]
  fn breaking() {
    let commit =
      ConventionalCommit::parse("123", r#"feat(core)!: implement `my-module` crate"#).unwrap();
    assert!(commit.is_breaking);
    let commit = ConventionalCommit::parse(
      "123",
      r#"feat(core): implement `my-module` crate

BREAKING CHANGE: remove '.addSomething()' method.
"#,
    )
    .unwrap();
    assert!(commit.is_breaking);
  }

  #[test]
  fn parse_fail() {
    assert!(ConventionalCommit::parse("123", "any commit message").is_err());
    assert!(ConventionalCommit::parse("123", "unknown: any commit message").is_err());
  }
}

File: xtask/conventional-commit.ts
import type { Commit } from 'es-git';

const CONVENTIONAL_COMMIT_TYPES = [
  'feat',
  'fix',
  'refactor',
  'perf',
  'test',
  'style',
  'doc',
  'docs',
  'build',
  'ops',
] as const;

export type ConventionalCommitType = (typeof CONVENTIONAL_COMMIT_TYPES)[number];

export function isConventionalCommitType(x: unknown): x is ConventionalCommitType {
  return CONVENTIONAL_COMMIT_TYPES.some(type => type === x);
}

export class ConventionalCommit {
  public readonly sha: string;
  public readonly type: ConventionalCommitType;
  public readonly scopes: string[];
  public readonly isBreaking: boolean;
  public readonly summary: string;
  public readonly body?: string;

  static parse(sha: string, message: string): ConventionalCommit {
    const lines = message.split('\n');
    const [title = '', _, ...bodyLines] = lines;
    const m = /^(?<type>\w+)(?:\((?<scopes>.+)\))?(?<breaking>!)?: (?<summary>.+)$/.exec(title);
    const { type, scopes, breaking, summary } = m?.groups ?? {};
    if (!isConventionalCommitType(type)) {
      throw new Error(`invalid commit type: ${type}`);
    }
    if (summary == null) {
      throw new Error(`invalid commit summary: ${summary}`);
    }
    const body = bodyLines.length === 0 ? undefined : bodyLines.join('\n');
    return new ConventionalCommit(
      sha,
      type,
      scopes?.split(',').map(x => x.trim()) ?? [],
      breaking != null || body?.includes('BREAKING CHANGE:') === true,
      summary,
      body
    );
  }

  static parseCommit(commit: Commit): ConventionalCommit {
    const sha = commit.id();
    const message = commit.message();
    return ConventionalCommit.parse(sha, message);
  }

  private constructor(
    sha: string,
    type: ConventionalCommitType,
    scopes: string[],
    isBreaking: boolean,
    summary: string,
    body?: string
  ) {
    this.sha = sha;
    this.type = type;
    this.scopes = scopes;
    this.isBreaking = isBreaking;
    this.summary = summary;
    this.body = body;
  }
}
